tools/content/unimplemented_katex_test.dart

```diff
@@ -0,0 +1,161 @@
+        final reason = 'hard fail: ${hardFailReason.error} "$firstLine"';
+        (failedMessageIdsByReason[reason] ??= {}).add(messageId);
+        (failedMathNodesByReason[reason] ??= []).add(value);
+        failureCount++;
+      }
+
+      if (softFailReason != null) {
+        for (final cssClass in softFailReason.unsupportedCssClasses) {
+          final reason = 'unsupported css class: $cssClass';
+          (failedMessageIdsByReason[reason] ??= {}).add(messageId);
+          (failedMathNodesByReason[reason] ??= []).add(value);
+          failureCount++;
+        }
+        for (final cssProp in softFailReason.unsupportedInlineCssProperties) {
+          final reason = 'unsupported inline css property: $cssProp';
+          (failedMessageIdsByReason[reason] ??= {}).add(messageId);
+          (failedMathNodesByReason[reason] ??= []).add(value);
+          failureCount++;
+        }
+      }
+
+      if (failureCount == 0) {
+        final reason = 'unknown';
+        (failedMessageIdsByReason[reason] ??= {}).add(messageId);
+        (failedMathNodesByReason[reason] ??= []).add(value);
+      }
+    }
+
+    await for (final message in readMessagesFromJsonl(file)) {
+      totalMessageCount++;
+      walk(message.id, parseContent(message.content).toDiagnosticsNode());
+    }
+
+    final buf = StringBuffer();
+    buf.writeln();
+    buf.writeln('Out of $totalMessageCount total messages,'
+      ' ${katexMessageIds.length} of them were KaTeX containing messages'
+      ' and ${failedKatexMessageIds.length} of them failed.');
+    buf.writeln('There were $totalMathBlockNodes math block nodes out of which $failedMathBlockNodes failed.');
+    buf.writeln('There were $totalMathInlineNodes math inline nodes out of which $failedMathInlineNodes failed.');
+    buf.writeln();
+
+    for (final MapEntry(key: reason, value: messageIds) in failedMessageIdsByReason.entries.sorted(
+      (a, b) => b.value.length.compareTo(a.value.length),
+    )) {
+      final failedMathNodes = failedMathNodesByReason[reason]!;
+      int oldestId = messageIds.reduce(min);
+      int newestId = messageIds.reduce(max);
+
+      buf.writeln('Because of $reason:');
+      buf.writeln('  ${messageIds.length} messages failed.');
+      buf.writeln('  Oldest message: $oldestId, Newest message: $newestId');
+      if (!_verbose) {
+        buf.writeln();
+        continue;
+      }
+
+      buf.writeln('  Message IDs (upto 100): ${messageIds.take(100).join(', ')}');
+      buf.writeln('  TeX source (upto 30):');
+      for (final node in failedMathNodes.take(30)) {
+        final type = switch (node) {
+          MathBlockNode() => 'block',
+          MathInlineNode() => 'inline',
+        };
+        buf.writeln('    $type: "${node.texSource}"');
+      }
+      buf.writeln('  HTML (upto 10):');
+      for (final node in failedMathNodes.take(10)) {
+        buf.writeln('    ${node.debugHtmlText}');
+        buf.writeln();
+      }
```

Comment:
This seems to end up with a blank line separating the examples. Maybe skip the extra `writeln`?